### ISSUE TITLE
[no-unused-var] false positive when using recursive interface in d.ts

### ISSUE THREAD

**Issue by lakerswgq** [typescript-eslint/typescript-eslint#2714]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unused-vars": "error"
  }
}
```

interface.d.ts
```TS
// good
interface IPerson {
  name: string;
  age: number;
}
// got warning
interface IItem {
  title: string;
  url: string;
  children?: IItem[];
}
```

**Expected Result**

No warning on interface IItem

**Actual Result**
Got warning: 'IItem' is defined but never used

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.6.0` |
| `@typescript-eslint/parser`        | `4.6.0` |
| `TypeScript`                       | `4.0.5` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `12.16.2` |

**Comment by bradzacher** [comment]
are you using `IItem` elsewhere?
If no - then this is not a false-positive.
Self referencing things are not considered a usage - or else you'll get isolated, unused islands of things.

This is the [same behaviour that you get for self-referencing functions](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tdW51c2VkLXZhcnM6IFtcImVycm9yXCJdICovXG5cbmZ1bmN0aW9uIGZvbygpIHtcbiAgZm9vKCk7XG59XG4iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjExLCJzb3VyY2VUeXBlIjoibW9kdWxlIiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7fSwiZW52Ijp7fX19)

**Comment by lakerswgq** [comment]
@bradzacher yes，I declare `IItem` and some other interface as global interface in definition file，so i can access them in other files in my project.But only the `IItem` interface got warnings.

**Comment by bradzacher** [comment]
please provide the full code sample so I can repro it appropriately.

**Comment by lakerswgq** [comment]
@bradzacher try [this repo](https://github.com/lakerswgq/ts-test), please

**Comment by bradzacher** [comment]
The bug is that we're doing an explicit report for interfaces that self-reference, but that circumvents the checks we do for ambient declarations.

https://github.com/typescript-eslint/typescript-eslint/blob/0d696c72c5c9c3446902a63509d499ee95483e81/packages/eslint-plugin/src/rules/no-unused-vars.ts#L190-L199

**Comment by mateusppereira** [comment]
@bradzacher would this solve the issue?
```diff
        const isOnlySelfReferenced = variable.references.every(ref => {
          if (
            ref.identifier.range[0] >= node.range[0] &&
-             ref.identifier.range[1] <= node.range[1]
+            ref.identifier.range[1] <= node.range[1] &&
+            !util.isDefinitionFile(filename)
          ) {
            return true;
          }
          return false;
        });

```

**Comment by bradzacher** [comment]
that's in the right direction! But we could simplify this.

When the rule runs over a declaration file, it calls `markDeclarationChildAsUsed` to manually mark the variables (using a standard ESLint property - `eslintUsed`).

https://github.com/typescript-eslint/typescript-eslint/blob/0d696c72c5c9c3446902a63509d499ee95483e81/packages/eslint-plugin/src/rules/no-unused-vars.ts#L306-L358

So instead we could express this clearer as:
```diff
 const isOnlySelfReferenced = variable.references.every(ref => { 
   if ( 
     ref.identifier.range[0] >= node.range[0] && 
     ref.identifier.range[1] <= node.range[1] 
   ) { 
     return true; 
   } 
   return false; 
 }); 
-if (isOnlySelfReferenced) {
+if (isOnlySelfReferenced && variable.eslintUsed !== true) {
```

**Comment by ddubrava** [comment]
@bradzacher is it possible to create a test case for this? To test `variable.eslintUsed` we need two files: `.d.ts` and `.ts`.

**Comment by bradzacher** [comment]
yes - you can use the `filename` prop in the test case.
